crates/accelerate/src/two_qubit_decompose.rs

```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+            } else {
+                rand_a = state.sample(StandardNormal);
+                rand_b = state.sample(StandardNormal);
+            }
+            let m2_real = m2.mapv(|val| rand_a * val.re + rand_b * val.im).to_owned();
+            let p_inner = m2_real
+                .view()
+                .into_faer()
+                .selfadjoint_eigendecomposition(Lower)
+                .u()
+                .into_ndarray()
+                .mapv(Complex64::from)
+                .to_owned();
+            // Uncomment this to use numpy for eigh instead of faer (useful if needed to compare)
+            // let numpy_linalg = PyModule::import(py, "numpy.linalg")?;
+            // let eigh = numpy_linalg.getattr("eigh")?;
+            // let m2_real_arr = m2_real.to_pyarray(py);
+            // let result = eigh.call1((m2_real_arr,))?.downcast::<PyTuple>()?;
+            // let p_raw = result.get_item(1)?;
+            // let p_inner = p_raw
+            //     .extract::<PyReadonlyArray2<f64>>()?
+            //     .as_array()
+            //     .mapv(Complex64::from);
+            let d_inner = p_inner.t().dot(&m2).dot(&p_inner).diag().to_owned();
+            let mut diag_d: Array2<Complex64> = Array2::zeros((4, 4));
+            diag_d
+                .diag_mut()
+                .iter_mut()
+                .enumerate()
+                .for_each(|(index, x)| *x = d_inner[index]);
+
+            let compare = p_inner.dot(&diag_d).dot(&p_inner.t()).to_owned();
+            found = abs_diff_eq!(compare.view(), m2, epsilon = 1.0e-13);
+            if found {
+                p = p_inner;
+                d = d_inner;
+                break;
+            }
+        }
+        if !found {
+            import_exception!(qiskit, QiskitError);
+            return Err(QiskitError::new_err(format!(
+                "TwoQubitWeylDecomposition: failed to diagonalize M2. Please report this at https://github.com/Qiskit/qiskit-terra/issues/4159. Input: {:?}", unitary_matrix
+            )));
+        }
+        let mut d = -d.map(|x| x.arg() / 2.);
+        d[3] = -d[0] - d[1] - d[2];
+        let mut cs: Array1<f64> = (0..3)
+            .map(|i| ((d[i] + d[3]) / 2.0).rem_euclid(TWO_PI))
+            .collect();
+        let cstemp: Vec<f64> = cs
+            .iter()
+            .map(|x| x.rem_euclid(PI2))
+            .map(|x| x.min(PI2 - x))
+            .collect();
+        let mut order = utils::arg_sort(&cstemp);
+        (order[0], order[1], order[2]) = (order[1], order[2], order[0]);
+        (cs[0], cs[1], cs[2]) = (cs[order[0]], cs[order[1]], cs[order[2]]);
+        (d[0], d[1], d[2]) = (d[order[0]], d[order[1]], d[order[2]]);
+        let mut p_orig = p.clone();
+        for (i, item) in order.iter().enumerate().take(3) {
+            let slice_a = p.slice_mut(s![.., i]);
+            let slice_b = p_orig.slice_mut(s![.., *item]);
+            Zip::from(slice_a).and(slice_b).for_each(::std::mem::swap);
+        }
```

Comment:
At the time I couldn't find a pattern for doing this and some issues I dug up where people were asking for this suggested a similar pattern to this. But that might have been an old issue because it looks like I can probably do something like: `p.slice_mut(s![.., i]).assign(p.slice(s![..., *item])` (which internally is just doing the same clone and zip basically)

I seem to remember having trouble using `order` in the slice `s!` macro syntax though, so I might keep the for loop. Although that would result in three allocations for the internal clone in `assign()`  instead of just the one outer one.